src/cli/NativeMsgManager.cpp

```diff
@@ -0,0 +1,115 @@
+        {"type", "stdio"},
+        {"path", QCoreApplication::applicationFilePath()},
+    });
+
+    enum ManifestVariant {
+        Firefox,
+        Chromium
+    };
+
+    QMap<ManifestVariant, QJsonObject> manifests;
+
+    // For Firefox-based browsers (they do not like seeing "allowed_origins")
+    // only add the ones that match Firefox' schema definition. Otherwise it will
+    // fail to parse the manifest JSON.
+    // See https://searchfox.org/mozilla-central/rev/3419858c997f422e3e70020a46baae7f0ec6dacc/toolkit/components/extensions/schemas/manifest.json#571-583
+    QStringList extensions;
+    for (QString const &extension : nativeMessagingClients.values())
+        if (firefoxExtensionPattern.match(extension).hasMatch())
+            extensions << extension;
+
+    manifests.insert(Firefox, manifest);
+    manifests[Firefox]["allowed_extensions"] = QJsonArray::fromStringList(extensions);
+
+    // Chromium-based browsers look for full url origins
+    QStringList origins;
+    for (QString const &extension : nativeMessagingClients.values())
+        origins << QString("chrome-extension://%1/").arg(extension);
+
+    manifests.insert(Chromium, manifest);
+    manifests[Chromium]["allowed_origins"] = QJsonArray::fromStringList(origins);
+
+    QList<QPair<ManifestVariant,QString>> manifestPaths;
+
+#if defined(Q_OS_MACOS)
+    manifestPaths << qMakePair(Firefox,  QString("%1/Mozilla/NativeMessagingHosts/%2.json").arg(QStandardPaths::writableLocation(QStandardPaths::GenericDataLocation)).arg(name));
+    manifestPaths << qMakePair(Chromium, QString("%1/Google/Chrome/NativeMessagingHosts/%2.json").arg(QStandardPaths::writableLocation(QStandardPaths::GenericDataLocation)).arg(name));
+    manifestPaths << qMakePair(Chromium, QString("%1/Chromium/NativeMessagingHosts/%2.json").arg(QStandardPaths::writableLocation(QStandardPaths::GenericDataLocation)).arg(name));
+#elif defined (Q_OS_LINUX)
+    manifestPaths << qMakePair(Firefox,  QString("%1/.mozilla/native-messaging-hosts/%2.json").arg(QStandardPaths::writableLocation(QStandardPaths::HomeLocation)).arg(name));
+    manifestPaths << qMakePair(Chromium, QString("%1/.config/google-chrome/NativeMessagingHosts/%2.json").arg(QStandardPaths::writableLocation(QStandardPaths::HomeLocation)).arg(name));
+    manifestPaths << qMakePair(Chromium, QString("%1/.config/chromium/NativeMessagingHosts/%2.json").arg(QStandardPaths::writableLocation(QStandardPaths::HomeLocation)).arg(name));
+#elif defined (Q_OS_WIN)
+    // On Windows, we write the manifest to some safe directory, and then point to it from the Registry.
+    manifestPaths << qMakePair(Firefox,  QString("%1/%2-firefox.json").arg(QStandardPaths::writableLocation(QStandardPaths::AppConfigLocation)).arg(name));
+    manifestPaths << qMakePair(Chromium, QString("%1/%2-chromium.json").arg(QStandardPaths::writableLocation(QStandardPaths::AppConfigLocation)).arg(name));
```

Comment:
No chrome manifest for windows?

Reply:
Chromium and Chrome share the same file (but different registry keys, mentioned later on in the file.)